synapse/storage/databases/state/store.py

```diff
@@ -269,30 +428,14 @@ async def _get_state_for_groups(
         if not incomplete_groups:
             return state
 
-        cache_sequence_nm = self._state_group_cache.sequence
-        cache_sequence_m = self._state_group_members_cache.sequence
-
-        # Help the cache hit ratio by expanding the filter a bit
-        db_state_filter = state_filter.return_expanded()
-
-        group_to_state_dict = await self._get_state_groups_from_groups(
-            list(incomplete_groups), state_filter=db_state_filter
+        results_from_requests = await yieldable_gather_results(
+            self._get_state_for_group_using_inflight_cache,
+            incomplete_groups,
+            state_filter,
         )
 
-        # Now lets update the caches
-        self._insert_into_cache(
-            group_to_state_dict,
-            db_state_filter,
-            cache_seq_num_members=cache_sequence_m,
-            cache_seq_num_non_members=cache_sequence_nm,
-        )
-
-        # And finally update the result dict, by filtering out any extra
-        # stuff we pulled out of the database.
-        for group, group_state_dict in group_to_state_dict.items():
-            # We just replace any existing entries, as we will have loaded
-            # everything we need from the database anyway.
-            state[group] = state_filter.filter_state(group_state_dict)
+        for group, group_result in zip(incomplete_groups, results_from_requests):
```

Comment:
`incomplete_groups` seems to be a `set`, which the docs say "support partial ordering". I suspect iterating the same `set` multiple times would give the same order, but I don't know if we should trust that?

Reply:
Well spotted. Practically there would be very little reason I can see why a set should change iteration order between iterations, but it's not great practice since a mutation can resize the underlying hashtable, causing a rehash which can shuffle things around (at least in most languages. I know Python has guarantees about the order of dict keys, but don't see the same about sets.

> `set`, which the docs say "support partial ordering".

Not that I can see the passage in the documentation that you're referring to, but I suspect this is a red herring and is actually talking about a partial ordering on sets themselves. It'd be called a 'partial ordering' because if you have two sets A and B, you might find that neither `A < B` nor `B < A` holds.

The partial ordering, `<`, seems to mean 'is a subset of' here.
```python
>>> { 4 } < { 4, 8 }
True
>>> { 4, 7 } < { 4, 8 }
False
>>> { 4, 7 } > { 4, 8 }
False
```

Not super relevant but just a bit of trivia I suppose :)

Reply:
I think you're quite right, but I couldn't find any guarantee of ordering of sets. Practically though I couldn't make it go unordered...

Reply:
I'd prefer it if we didn't rely on that, especially as the consequences of getting it wrong is going to be horrific. Can we instead do something like:

```python
async def get_from_cache(group, filter):
    state[group] = await self._get_state_for_group_using_inflight_cache(group, filter)

await yieldable_gather_results(
    get_from_cache,
    incomplete_groups,
    state_filter,
)
```

Reply:
> I'd prefer it if we didn't rely on that, especially as the consequences of getting it wrong is going to be horrific. Can we instead do something like:

This was already changed before your comment, but I think I like your idea more...

Reply:
OH woops, clearly I can't read code :)